### PR TITLE
[POAE7-2540] PlanNode validation for cider plan validator

### PR DESCRIPTION
**Pull request by yma11**
### What changes were proposed in this pull request?
Add plan validator module for cider, with single plan node validation as first step

### Why are the changes needed?
provide plan validation capability for Cider

### Does this PR introduce _any_ user-facing change?
Yes.
Public APIs:
 static bool isSupported(const substrait::Plan& plan, const std::string& from_platform);
 static PlanSlice getCiderSupportedSlice(substrait::Plan& plan, std::string& from_platform);

### How was this patch tested?
UT

### CODE REVIEW

**Comment by yma11** [comment]
@winningsix @jikunshang @haiweizh Please help take a review. Thanks.

File: cider/exec/plan/validator/SingleNodeValidator.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+bool SingleNodeValidator::validate(const substrait::Rel& rel_node) {
+  // Validation on each plan node includes data types check and some basic
+  // properties check. Data types contains current node output types and its
+  // parents' output types. Actually there should be other data types involved
+  // in nested functions, which we don't cover here. This won't cause problem
+  // since it will be checked in later function level validation.
+  const substrait::Rel::RelTypeCase& rel_type = rel_node.rel_type_case();
+  switch (rel_type) {
+    case substrait::Rel::RelTypeCase::kRead:
+      return true;
```
Comment (haiweizh) [inline cider/exec/plan/validator/SingleNodeValidator.cpp:179]: use validate function

File: cider/exec/plan/validator/SingleNodeValidator.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+             isSupportedAllTypes(getRelOutputTypes(rel_node.join().right()));
+    }
+    default:
+      CIDER_THROW(CiderCompileException,
+                  fmt::format("Unsupported substrait rel type {}", rel_type));
+  }
+}
+
+bool SingleNodeValidator::validate(const substrait::ReadRel& read_rel) {
+  return true;
```
Comment (haiweizh) [inline cider/exec/plan/validator/SingleNodeValidator.cpp:207]: Why just return true?

File: cider/exec/plan/validator/SingleNodeValidator.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+bool SingleNodeValidator::validate(const substrait::JoinRel& join_rel) {
+  // Only support join with direct. For emit, it uses join with post project
+  return join_rel.has_expression() && join_rel.common().has_direct() &&
+         join_rel.type() !=
+             substrait::JoinRel_JoinType::JoinRel_JoinType_JOIN_TYPE_UNSPECIFIED &&
+         join_rel.type() !=
+             substrait::JoinRel_JoinType::JoinRel_JoinType_JOIN_TYPE_OUTER &&
+         join_rel.type() != substrait::JoinRel_JoinType::JoinRel_JoinType_JOIN_TYPE_RIGHT;
+}
+}  // namespace validator
```
Comment (haiweizh) [inline cider/exec/plan/validator/SingleNodeValidator.cpp:241]: Add line break.

File: cider/tests/plan/CiderPlanValidatorTest.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace bf = boost::filesystem;
+
+std::string getDataFilesPath() {
+  const std::string absolute_path = __FILE__;
+  auto const pos = absolute_path.find_last_of('/');
+  return absolute_path.substr(0, pos) + "/../substrait_plan_files/";
+}
+
+TEST(CiderPlanValidator, InvalidAggAndJoinTest) {
```
Comment (haiweizh) [inline cider/tests/plan/CiderPlanValidatorTest.cpp:37]: Should add more test case?

File: cider/exec/plan/validator/CiderPlanValidator.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  // Call isSupportedSlice() to do further validation on plan slices
+  // return the longest plan slice
+  int final_slice_index = 0, max_len = 0;
+  for (int i = 0; i < slice_candidates.size(); i++) {
+    if (slice_candidates[i].rel_nodes.size() > max_len) {
+      max_len = slice_candidates[i].rel_nodes.size();
+      final_slice_index = i;
+    }
+  }
+  return slice_candidates[final_slice_index];
```
Comment (haiweizh) [inline cider/exec/plan/validator/CiderPlanValidator.cpp:132]: Will there be scenes that need to be split into multiple？

File: cider/exec/plan/validator/SingleNodeValidator.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include "SingleNodeValidator.h"
```
Comment (harborn) [inline cider/exec/plan/validator/SingleNodeValidator.cpp:24]: I think here should be:
`#include "cider/exec/plan/validator/SingleNodeValidator.h"`

File: cider/exec/plan/validator/CiderPlanValidator.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+PlanSlice CiderPlanValidator::getCiderSupportedSlice(substrait::Plan& plan,
+                                                     std::string& from_platform) {
+  substrait::Rel root = getRootRel(plan);
+  // Put plan rels in a vector with pair <substrait::Rel, isSupported>
+  std::vector<substrait::Rel> rel_vec;
+  putRelNodesInVec(root, rel_vec);
+  // Compose plan slices based on SingleNodeValidator validate result
+  std::vector<PlanSlice> slice_candidates;
+  for (int i = 0; i < rel_vec.size(); i++) {
```
Comment (jikunshang) [inline cider/exec/plan/validator/CiderPlanValidator.cpp:106]: after this loop it's possible to have 2 or more sequence slice. if we have a plan like `A->B->C->D->E->F->G`,  it's possible to have `[B->C]`,  `[E->F]` in slice_candidates.

File: cider/exec/plan/validator/CiderPlanValidator.cpp (hunk 5 above)
Comment (jikunshang) [inline cider/exec/plan/validator/CiderPlanValidator.cpp:132]: not safe. it's possible to be empty vector.

Comment (yma11) [reply cider/exec/plan/validator/CiderPlanValidator.cpp:132]: For a PlanSlice, we won't split again.

File: cider/exec/plan/validator/SingleNodeValidator.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+                  fmt::format("Unsupported substrait rel type {}", rel_type));
+  }
+}
+
+bool SingleNodeValidator::validate(const substrait::ReadRel& read_rel) {
+  return true;
+}
+
+bool SingleNodeValidator::validate(const substrait::FilterRel& filter_rel) {
+  return filter_rel.has_condition();
```
[the remark this answers is not in the document]
Comment (yma11) [reply cider/exec/plan/validator/SingleNodeValidator.cpp:211]: Yes. Single plan node validation is the first step, in order to split the input plan to some plan slices. Function look up will happen in next step.

File: cider/exec/plan/validator/CiderPlanValidator.cpp (hunk 7 above)
Comment (yma11) [reply cider/exec/plan/validator/CiderPlanValidator.cpp:106]: Yes. If they have same length, we will return the first one.  We may can add better rule for final slice select later.

File: cider/exec/plan/validator/CiderPlanValidator.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  // Validate plan slice by function lookup and rule based check
+  static bool isSupportedSlice(const PlanSlice& plan_slice,
+                               const std::string& from_platform);
+
+  static substrait::Rel getRootRel(const substrait::Plan& plan);
+
+  static PlanSlice constructPlanSlice(const substrait::Plan& plan,
+                                      const std::string& from_platform);
+
+  static void putRelNodesInVec(const substrait::Rel& rel_node,
```
[the remark this answers is not in the document]
Comment (yma11) [reply cider/exec/plan/validator/CiderPlanValidator.h:62]: It's actually not a flatten process but just put them in a vector with root to source manner.

File: cider/exec/plan/validator/SingleNodeValidator.cpp [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+  // parents' output types. Actually there should be other data types involved
+  // in nested functions, which we don't cover here. This won't cause problem
+  // since it will be checked in later function level validation.
+  const substrait::Rel::RelTypeCase& rel_type = rel_node.rel_type_case();
+  switch (rel_type) {
+    case substrait::Rel::RelTypeCase::kRead:
+      return true;
+    case substrait::Rel::RelTypeCase::kFilter: {
+      return validate(rel_node.filter()) &&
+             isSupportedAllTypes(getRelOutputTypes(rel_node.filter().input()));
```
[the remark this answers is not in the document]
Comment (yma11) [reply cider/exec/plan/validator/SingleNodeValidator.cpp:182]: It's okay to put it here or inside the `validate` method. For easy implementation to cover different rel node types, I put it here.

File: cider/tests/plan/CiderPlanValidatorTest.cpp (hunk 4 above)
Comment (yma11) [reply cider/tests/plan/CiderPlanValidatorTest.cpp:37]: will add more in next PR when trying to generate EU and do function lookup.

File: cider/exec/plan/validator/SingleNodeValidator.cpp (hunk 2 above)
Comment (yma11) [reply cider/exec/plan/validator/SingleNodeValidator.cpp]: updated.

File: cider/exec/plan/validator/SingleNodeValidator.cpp (hunk 3 above)
Comment (yma11) [reply cider/exec/plan/validator/SingleNodeValidator.cpp:248]: Done.

File: cider/exec/plan/validator/SingleNodeValidator.cpp (hunk 6 above)
Comment (yma11) [reply cider/exec/plan/validator/SingleNodeValidator.cpp:24]: This header file is in same folder as this same cpp file so it should be okay to put here?

File: cider/exec/plan/validator/CiderPlanValidator.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "substrait/algebra.pb.h"
+#include "substrait/plan.pb.h"
+#include "substrait/type.pb.h"
+
+namespace validator {
+
+struct PlanSlice {
+  std::vector<substrait::Rel> rel_nodes;
+  std::string from_platform;
```
[the remark this answers is not in the document]
Comment (yma11) [reply cider/exec/plan/validator/CiderPlanValidator.h]: updated.

File: cider/exec/plan/validator/CiderPlanValidator.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+class CiderPlanValidator {
+ public:
+  /**
+   * @param plan plan with pattern already validated
+   * @param from_platform specific frontend framework for future function lookup
+   *
+   **/
+  static bool isSupported(const substrait::Plan& plan, const std::string& from_platform);
```
[the remark this answers is not in the document]
Comment (yma11) [reply cider/exec/plan/validator/CiderPlanValidator.h]: updated.

File: cider/exec/plan/validator/CiderPlanValidator.cpp [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+      putRelNodesInVec(rel_node.join().left(), rel_vec);
+      break;
+    }
+    default:
+      CIDER_THROW(CiderCompileException,
+                  fmt::format("Unsupported substrait rel type {}", rel_type));
+  }
+}
+
+substrait::Rel CiderPlanValidator::getRootRel(const substrait::Plan& plan) {
```
[the remark this answers is not in the document]
Comment (yma11) [reply cider/exec/plan/validator/CiderPlanValidator.cpp]: updated.

**Review by jikunshang — APPROVED** [review]

**Review by haiweizh — APPROVED** [review]

**Review by spevenhe — APPROVED** [review]

**Pull request merged**